### PR TITLE
Zingoledger traits

### PR DESCRIPTION
**Pull request by fluidvanadium**

### CODE REVIEW

File: zingolib/src/wallet/notes/interface.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    keys::unified::WalletCapability,
+    traits::{FromBytes, FromCommitment, Nullifier, ReadableWriteable, ToBytes},
+    Pool,
+};
+
+pub trait NoteInterface: Sized {
+    fn spent(&self) -> &Option<(TxId, u32)>;
+    fn is_spent(&self) -> bool {
+        Self::spent(self).is_some()
+    }
```
Comment (Oscar-Pepper) [inline zingolib/src/wallet/notes/interface.rs:15]: it makes logical sense to me to have is_spent as the first or last fn so spent and spent_mut etc. are grouped together. just an organisation thing

File: zingolib/src/wallet/notes/interface.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    fn spent_mut(&mut self) -> &mut Option<(TxId, u32)>;
+    fn pending_spent(&self) -> &Option<(TxId, u32)>;
+    fn pending_spent_mut(&mut self) -> &mut Option<(TxId, u32)>;
+}
+
+///   All zingolib::wallet::traits::Notes are NoteInterface
+///   NoteInterface provides...
```
Comment (Oscar-Pepper) [inline zingolib/src/wallet/notes/interface.rs]: this doc comment is for NoteInterface not ShieldedNoteInterface.... also its not very helpful haha

i think everything in wallet::traits is for shielded right? or maybe the from bytes/ to bytes might be applicable to transprent im not sure

Comment (Oscar-Pepper) [reply zingolib/src/wallet/notes/interface.rs]: i see this comment was made before you created NoteInterface... interesting..

maybe change to ShieldedNoteInterface?

File: zingolib/src/wallet/notes/transparent.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            address,
+            txid: transaction_id,
+            output_index,
+            script,
+            value,
+            spent: spent_tuple,
+            unconfirmed_spent: None,
+        })
+    }
+}
```
Comment (Oscar-Pepper) [inline zingolib/src/wallet/notes/transparent.rs:167]: not sure why read and write were swapped in order but not an issue, just harder to review and dont see the benefit

File: zingolib/src/wallet/notes/transparent.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    fn pending_spent(&self) -> &Option<(TxId, u32)> {
+        &self.unconfirmed_spent
+    }
+    fn pending_spent_mut(&mut self) -> &mut Option<(TxId, u32)> {
+        &mut self.unconfirmed_spent
+    }
+}
+
+impl TransparentNote {
+    pub fn new(
```
Comment (Oscar-Pepper) [inline zingolib/src/wallet/notes/transparent.rs]: can we change this to `from_parts` to stay in line with sapling/orchard and also the LRZ format?

**Review by Oscar-Pepper — CHANGES_REQUESTED** [review]
this is really great work, I have a  few questions/requests.
i noticed after running a global search on `transaction_meta` that there are many places where transaction_metadata is still used instead of transaction_record and also transaction_metadata_set is used instead of zingo_ledger. could you either fix this or put an issue out with the merging of this PR so its pulled in line soon?
also there is a mistake in wallet/traits.rs:399 where is says `transaction_metadata_set` but its a `transaction_metadata`

File: zingolib/src/wallet/notes/transparent.rs (hunk 3 above)
Comment (zancas) [reply zingolib/src/wallet/notes/transparent.rs:167]: If this was the result of an automation we should configure the tool to enforce the preferred order.

File: zingolib/src/wallet/notes/interface.rs (hunk 2 above)
Comment (fluidvanadium) [reply zingolib/src/wallet/notes/interface.rs]: fixed in https://github.com/zingolabs/zingolib/pull/809

File: zingolib/src/wallet/notes/transparent.rs (hunk 3 above)
Comment (fluidvanadium) [reply zingolib/src/wallet/notes/transparent.rs:167]: sorry for any confusion. i feel strongly that write comes before read in a flow

**Pull request closed without merging**